src/simulator/halo_simulator.py

```diff
@@ -235,6 +235,50 @@ def simulated_venn_diagram_reach_by_spend(
         """
         raise NotImplementedError()
 
+    def _generate_reach_points_from_venn_diagram(
+        self, spends: List[float], primitive_regions: Dict[int, List]
+    ) -> List[ReachPoint]:
+        """Return the reach points of the powerset of active publishers.
+
+        For each subset of active publishers, compute reach and frequency
+        estimate for those users who are reached by at least one of the
+        publishers in the subset.
+
+        Args:
+            spends:  The hypothetical spend vector, equal in length to
+              the number of publishers.  spends[i] is the amount that is
+              spent with publisher i.
+            primitive_regions:  Contains k+ reaches in the regions. The k+
+              reaches for a given region is given as a list r[] where r[k] is
+              the number of people who were reached AT LEAST k+1 times.
+        Returns:
+            A list of ReachPoint. Each reach point represents the mapping from
+            the spends of a subset of publishers to the number of people reached
+            in this subset.
+        """
+        active_pub_set = [i for i in range(len(spends)) if spends[i]]
+        active_pub_powerset = chain.from_iterable(
+            combinations(active_pub_set, r) for r in range(1, len(active_pub_set) + 1)
+        )
+        impressions = self._data_set.impressions_by_spend(spends)
+
+        reach_points = []
+        for sub_pub_ids in active_pub_powerset:
+            sub_reach = self._aggregate_reach_in_primitive_venn_diagram_regions(
+                sub_pub_ids, primitive_regions
+            )
+            pub_subset = set(sub_pub_ids)
+            sub_imps = [
+                imps if pub_id in pub_subset else 0
```

Comment:
It's slightly more compact to write this as
```
  imps for pub_id, imps in enumerate(impressions) if pub_id in pub_subset
```

Maybe an even more compact approach would be as follows:
```
  pub_vector = np.array([int(i in sub_pub_ids) for i in range(len(spends))])
  sub_imps = np.array(imps) * pub_vector
  sub_spends = np.array(spends) * pub_vector
```

Reply:
Thanks for the suggestions! I like the second approach and will change the code to that.
Just note that I think the first approach is not correct because it will generate a list with only non-zero values. For example, 
```python
pub_subset = set([3, 5, 8])
impressions = list(range(10))
print([imps for pub_id, imps in enumerate(impressions) if pub_id in pub_subset])
>>> [3, 5, 8]
```
Instead of `[0 0 0 3 0 5 0 0 8 0]` which is the desired output.

Reply:
Based on the discussion with @matthewclegg , we will change the code to the second approach which is suggested above.

Reply:
Done.